File: src/nbla/function/generic/softmax.cpp
// Copyright (c) 2017 Sony Corporation. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// softmax.cpp

#include <nbla/array.hpp>
#include <nbla/function/softmax.hpp>
#include <nbla/variable.hpp>

#include <algorithm>
#include <cmath>

namespace nbla {

NBLA_REGISTER_FUNCTION_SOURCE(Softmax, int);

template <typename T>
void Softmax<T>::setup_impl(const Variables &inputs, const Variables &outputs) {
  Shape_t in_shape = inputs[0]->shape();
  NBLA_CHECK(axis_ < in_shape.size(), error_code::value,
             "axis must be less than ndim of inputs[0]. "
             "axis: %d >= ndim of inputs[0]: %d.",
             axis_, in_shape.size());
  outputs[0]->reshape(in_shape, true);
  Size_t size = inputs[0]->size();
  Size_t size_axis = inputs[0]->size(axis_);
  size0_ = size / size_axis;          // Batch size.
  size1_ = inputs[0]->shape()[axis_]; // Size of specified axis.
  size2_ = size / size0_ / size1_;    // Size of rest.
  NBLA_CHECK(size0_ * size1_ * size2_ == size, error_code::unclassified,
             "An error occurred during setup Softmax function.");
}

template <class T>
void Softmax<T>::forward_impl(const Variables &inputs,
                              const Variables &outputs) {
  typedef typename force_float<T>::type AccumType;
  // Setting up variables
  const T *x = inputs[0]->get_data_pointer<T>(this->ctx_);
  T *y = outputs[0]->cast_data_and_get_pointer<T>(this->ctx_, true);
  for (int i0 = 0; i0 < size0_; ++i0) {
    for (int i2 = 0; i2 < size2_; ++i2) {
      const int j = i0 * size1_ * size2_ + i2;
      // compute maximum
      T max_x = x[j];
      for (int i1 = 0; i1 < size1_; ++i1) {
        const int k = i1 * size2_ + j;
        max_x = (max_x >= x[k]) ? max_x : x[k];
      }
      // Compute exponential and sum
      AccumType exp_sum = 0;
      for (int i1 = 0; i1 < size1_; ++i1) {
        const int k = i1 * size2_ + j;
        const T tmp = std::exp(x[k] - max_x);
        y[k] = tmp;
        exp_sum += tmp;
      }
      // Compute softmax
      for (int i1 = 0; i1 < size1_; ++i1) {
        const int k = i1 * size2_ + j;
        y[k] = y[k] / exp_sum;
      }
    }
  }
}

template <class T>
void Softmax<T>::backward_impl(const Variables &inputs,
                               const Variables &outputs,
                               const vector<bool> &propagate_down,
                               const vector<bool> &accum) {
  if (!propagate_down[0]) {
    return;
  }
  typedef typename force_float<T>::type AccumType;
  // Setting up variables
  const T *y = outputs[0]->get_data_pointer<T>(this->ctx_);
  const T *dy = outputs[0]->get_grad_pointer<T>(this->ctx_);
  T *dx = inputs[0]->cast_grad_and_get_pointer<T>(this->ctx_, !accum[0]);

  for (int i0 = 0; i0 < size0_; ++i0) {
    for (int i2 = 0; i2 < size2_; ++i2) {
      const int j = i0 * size1_ * size2_ + i2;
      // compute sum of dy * y
      AccumType dyy_sum = 0;
      for (int i1 = 0; i1 < size1_; ++i1) {
        const int k = i1 * size2_ + j;
        dyy_sum += dy[k] * y[k];
      }
      // Compute backward
      for (int i1 = 0; i1 < size1_; ++i1) {
        const int k = i1 * size2_ + j;
        dx[k] = (accum[0] ? dx[k] : (T)0) + y[k] * (dy[k] - dyy_sum);
      }
    }
  }
}
}

File: python/src/nnabla/backward_function/softmax.py
# Copyright (c) 2017 Sony Corporation. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import nnabla as nn
import nnabla.functions as F
from .backward_function import BackwardFunction


class SoftmaxBackward(BackwardFunction):

    @property
    def name(self):
        return 'SoftmaxBackward'

    def _create_forward_inputs_and_outputs(self, inputs, outputs):
        # Inputs on the forward graph
        inputs_fwd = []
        for i in range(self._num_inputs_fwd):
            need_grad = self.forward_func.inputs[i].need_grad
            v = nn.Variable(inputs[i].shape, need_grad=need_grad)
            v.data = inputs[i].data
            v.grad = outputs[i].data
            inputs_fwd += [v]
        # Outputs on the forward graph
        outputs_fwd = []
        inp0 = inputs[self._num_inputs_fwd]      # y
        inp1 = inputs[self._num_inputs_fwd + 1]  # dy
        v = nn.Variable(inp1.shape)
        v.data = inp0.data
        v.grad = inp1.data
        outputs_fwd += [v]
        return inputs_fwd, outputs_fwd

    def backward_impl(self, inputs, outputs, prop_down, accum):
        # inputs: [inputs_fwd_graph] + [inputs_bwd_graph] or
        # [inputs_fwd_graph] + [outputs_fwd_graph] + [inputs_bwd_graph]

        axis = self.forward_func.info.args["axis"]

        # Inputs
        x0 = inputs[0].data
        y0 = inputs[1].data
        dy = inputs[2].data
        # Outputs
        dx0 = outputs[0].data
        # Grads of inputs
        g_x0 = inputs[0].grad
        g_y0 = inputs[1].grad
        g_dy = inputs[2].grad
        # Grads of outputs
        g_dx0 = outputs[0].grad

        # w.r.t. x0
        if prop_down[0]:
            gdx_y = g_dx0 * y0
            gdx_dy_y = gdx_y * dy
            dy_y = dy * y0
            gdx_y_sum = F.sum(gdx_y, axis, True)
            dy_y_sum = F.sum(dy_y, axis, True)
            gdx_dy_y_sum = F.sum(gdx_dy_y, axis, True)

            t1 = gdx_dy_y
            t2 = y0 * gdx_dy_y_sum
            t3 = gdx_y_sum * (y0 * dy_y_sum - dy_y)
            t4 = dy_y_sum * (y0 * gdx_y_sum - gdx_y)

            g_x0_ = t1 - t2 + t3 + t4
            if accum[0]:
                g_x0 += g_x0_
            else:
                g_x0.copy_from(g_x0_)

        # w.r.t. dy
        if prop_down[2]:
            si = nn.Variable(x0.shape).apply(
                data=x0, grad=g_dy, need_grad=True)
            so = nn.Variable(dx0.shape).apply(data=y0, grad=g_dx0)
            self.forward_func.backward([si], [so], accum=[accum[2]])
